### PR TITLE
Poset refactor part 2: lattices

### PR DESCRIPTION
**Pull request by plt-amy**
Opening a PR so I can use the review interface for these scattered comments.. We can change the base branch once #293 is merged.

### CODE REVIEW

File: src/Order/Frame.lagda.md [hunk 1]
```diff
@@ -18,140 +22,196 @@ module Order.Frame where
...
+        (lift true) → x
+        (lift false) → y
+    has-join-slat .is-join-semilattice.has-joins x y .Join.has-join .is-join.l≤join =
+      fam≤⋃ (lift true)
+    has-join-slat .is-join-semilattice.has-joins x y .Join.has-join .is-join.r≤join =
+      fam≤⋃ (lift false)
+    has-join-slat .is-join-semilattice.has-joins x y .Join.has-join .is-join.least ub' x≤ub' y≤ub' =
+      ⋃-universal _ λ where
+        (lift true) → x≤ub'
+        (lift false) → y≤ub'
```
Comment (plt-amy) [inline src/Order/Frame.lagda.md:67]: This should be factored out into properties of lubs (and we should have the corresponding theorem for glbs)

File: src/Order/Frame.lagda.md [hunk 2]
```diff
@@ -18,140 +22,196 @@ module Order.Frame where
...
+  open has-lubs renaming (fam≤lub to fam≤⋃; least to ⋃-universal) public
 
-```agda
-record is-frame
-  (⊤   : A)
-  (_∩_ : A → A → A)
-  (⋃ : ∀ {I : Type (level-of A)} → (I → A) → A)
-  : Type (lsuc (level-of A)) where
+  ⋃ : ∀ {I : Type o} → (I → ⌞ P ⌟) → ⌞ P ⌟
+  ⋃ k = has-lubs.lub {k = k}
```
Comment (plt-amy) [inline src/Order/Frame.lagda.md:39]: Maybe not as part of this PR, but it's worth noting that we need better API for joins of oversized families (i.e. what I did [here](https://github.com/plt-amy/1lab/blob/main/src/Cat/Displayed/Doctrine/Frame.lagda.md#L169))

File: src/Order/Frame.lagda.md [hunk 3]
```diff
@@ -18,140 +22,196 @@ module Order.Frame where
...
--->
-
-A **frame** is a lattice with finite meets^[So, in addition to the $x
-\cap y$ operation, we have a top element] and arbitrary joins satisfying
-the **infinite distributive law**
+record is-frame {o ℓ} (P : Poset o ℓ) : Type (lsuc o ⊔ ℓ) where
+  no-eta-equality
+  field
+    has-meet-slat : is-meet-semilattice P
+    has-lubs : ∀ {I : Type o} (k : I → ⌞ P ⌟) → Lub P k
```
Comment (plt-amy) [inline src/Order/Frame.lagda.md:30]: Align the `:`s; not sure about the ergonomics of nesting `is-meet-semilattice`, maybe it'd be better to have it inlined?

File: src/Order/Frame.lagda.md [hunk 4]
```diff
@@ -18,140 +22,196 @@ module Order.Frame where
...
+    module P = Poset P
+    module Pᶠ = is-frame P-frame
+    module Q = Order.Reasoning Q
+    module Qᶠ = is-frame Q-frame
+    open is-lub
   field
-    pres-⊤ : f X.top ≡ Y.top
-    pres-∩ : ∀ x y → f (x X.∩ y) ≡ (f x Y.∩ f y)
-    pres-⋃ : ∀ {I} (g : I → A) → f (X.⋃ g) ≡ Y.⋃ λ i → f (g i)
+    has-meet-slat-hom : is-meet-slat-hom f Pᶠ.has-meet-slat Qᶠ.has-meet-slat
```
Comment (plt-amy) [inline src/Order/Frame.lagda.md:119]: Whatever we decide regarding whether to inline `is-meet-semilattice` above should also apply to this field

File: src/Order/Frame.lagda.md [hunk 5]
```diff
@@ -172,8 +232,8 @@ has a join. This is a good definition for the **join of the subset
 $P$**.
 
 ```agda
-module _ {ℓ} (F : Frames.Ob ℓ) where
-  private module F = Frame-on (F .snd)
+module _ {o ℓ} (F : Frame o ℓ) where
```
Comment (plt-amy) [inline src/Order/Frame.lagda.md:235]: This stuff could probably get exploded into another module too

File: src/Order/Semilattice.lagda.md [hunk 6]
```diff
@@ -25,319 +21,384 @@ module Order.Semilattice where
...
+  ⋂ᶠ-≤ {n = suc n} (fsuc i) = ≤-trans ∩-≤r (⋂ᶠ-≤ i)
+
+  ⋂ᶠ-universal
+    : ∀ {n} {f : Fin n → ⌞ P ⌟}
+    → (x : ⌞ P ⌟)
+    → (∀ i → x ≤ f i)
+    → x ≤ ⋂ᶠ f
+  ⋂ᶠ-universal {n = zero} {f = f} x p = !
+  ⋂ᶠ-universal {n = suc n} {f = f} x p =
+    ∩-universal _ (p fzero) (⋂ᶠ-universal x (p ⊙ fsuc))
```
Comment (plt-amy) [inline src/Order/Semilattice.lagda.md:87]: This should end up in its own module about meet semilattices

File: src/Order/Semilattice.lagda.md [hunk 7]
```diff
@@ -25,319 +21,384 @@ module Order.Semilattice where
...
+  ⋃ᶠ-≤ {n = suc n} (fsuc i) = ≤-trans (⋃ᶠ-≤ i) ∪-≤r
+
+  ⋃ᶠ-universal
+    : ∀ {n} {f : Fin n → ⌞ P ⌟}
+    → (x : ⌞ P ⌟)
+    → (∀ i → f i ≤ x)
+    → ⋃ᶠ f ≤ x
+  ⋃ᶠ-universal {n = zero} {f = f} x p = ¡
+  ⋃ᶠ-universal {n = suc n} {f = f} x p =
+    ∪-universal _ (p fzero) (⋃ᶠ-universal x (p ⊙ fsuc))
```
Comment (plt-amy) [inline src/Order/Semilattice.lagda.md:154]: Ditto for join semilattices

File: src/Order/Semilattice.lagda.md [hunk 8]
```diff
@@ -25,319 +21,384 @@ module Order.Semilattice where
...
+    ∩-≤ : ∀ x y → (f # x) Qₗ.∩ (f # y) Q.≤ f # (x Pₗ.∩ y)
+    top-≤ : Qₗ.top Q.≤ f # Pₗ.top
+
+  pres-∩ : ∀ x y → f # (x Pₗ.∩ y) ≡ (f # x) Qₗ.∩ (f # y)
+  pres-∩ x y =
+    Q.≤-antisym
+      (Qₗ.∩-universal (f # (x Pₗ.∩ y))
+        (f .pres-≤ Pₗ.∩-≤l)
+        (f .pres-≤ Pₗ.∩-≤r))
+      (∩-≤ x y)
```
Comment (plt-amy) [inline src/Order/Semilattice.lagda.md:221]: Can we get preservation of finite caps too?

**Review by plt-amy — COMMENTED** [review]
Other than the implicit comments about exploding meet and join semilattices, there's a few other things to iron out IMO

File: src/Order/Semilattice.lagda.md (hunk 6 above)
Comment (plt-amy) [reply src/Order/Semilattice.lagda.md:87]: Also we can definitely make the arguments to the properties implicit right?

File: src/Cat/Displayed/Doctrine/Frame.lagda.md [hunk 9]
```diff
@@ -180,8 +181,8 @@ a calculation:
 ```agda
     lifted : Cocartesian-lift disp f g
     lifted .y' y      = exist y
-    lifted .lifting i = F.⋃-colimiting (g i , inc (i , refl , refl)) fst
-    lifted .cocartesian .universal {u' = u'} m h' i = F.⋃-universal fst λ (e , b) →
+    lifted .lifting i = F.fam≤⋃ (g i , inc (i , refl , refl))
```
Comment (plt-amy) [inline src/Cat/Displayed/Doctrine/Frame.lagda.md:184]: So maybe `⋃-colimiting` wasn't the best name, but I don't like this either; what about `⋃-coproj`?

File: src/Order/Frame.lagda.md (hunk 3 above)
Comment (TOTBWF) [reply src/Order/Frame.lagda.md:30]: Yeah, this is something I'm still pondering over. In practice we normally define things in one go as opposed to lots of little steps, so it's worth trying!

File: src/Order/Lattice.lagda.md [hunk 10]
```diff
@@ -187,146 +74,83 @@ finite joins.
...
 
   field
-    pres-⊤ : f S.top ≡ T.top
-    pres-⊥ : f S.bot ≡ T.bot
-    pres-∩ : ∀ {x y} → f (x S.∩ y) ≡ f x T.∩ f y
-    pres-∪ : ∀ {x y} → f (x S.∪ y) ≡ f x T.∪ f y
+    top-≤ : T.top Q.≤ f # S.top
+    bot-≤ : f # S.bot Q.≤ T.bot
+    ∩-≤ : ∀ {x y} → f # x T.∩ f # y Q.≤ f # (x S.∩ y)
+    ∪-≤ : ∀ {x y} → f # (x S.∪ y) Q.≤ f # x T.∪ f # y
```
Comment (plt-amy) [inline src/Order/Lattice.lagda.md:94]: wouldn't it be easier to have these as equations?

Comment (TOTBWF) [reply src/Order/Lattice.lagda.md:94]: When you actually go to prove these, one half of the proof is always the universal property of meets/joins: it's a bit easier to have  the equations as a theorem attached to the record.

Comment (plt-amy) [reply src/Order/Lattice.lagda.md:94]: sensible

**Comment by Lavenza** [comment]
[Pull request preview](https://preview.1lab.dev/295)

**Review by Lavenza — APPROVED** [review]

**Pull request merged**